listings/applications/erc721/src/erc721.cairo

```diff
@@ -0,0 +1,209 @@
+        from: ContractAddress,
+        to: ContractAddress,
+        token_id: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ApprovalForAll {
+        owner: ContractAddress,
+        operator: ContractAddress,
+        approved: bool
+    }
+
+    mod Errors {
+        // Error messages for each function
+        pub const ADDRESS_ZERO: felt252 = 'ERC721: invalid owner';
+        pub const INVALID_TOKEN_ID: felt252 = 'ERC721: invalid token ID';
+        pub const APPROVAL_TO_CURRENT_OWNER: felt252 = 'ERC721: approval to owner';
+        pub const NOT_TOKEN_OWNER: felt252 = 'ERC721:  not token owner';
+        pub const APPROVE_TO_CALLER: felt252 = 'ERC721: approve to caller';
+        pub const CALLER_NOT_OWNER_NOR_APPROVED: felt252 = 'ERC721: caller is not approved';
+        pub const CALLER_IS_NOT_OWNER: felt252 = 'ERC721: caller is not owner';
+        pub const TRANSFER_TO_ZERO_ADDRESS: felt252 = 'ERC721: to the zero address';
+        pub const TOKEN_ALREADY_MINTED: felt252 = 'ERC721: token already minted';
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, _name: felt252, _symbol: felt252) {
+        self.name.write(_name);
+        self.symbol.write(_symbol);
+    }
+
+    #[generate_trait]
+    impl IERC721Impl of IERC721Trait {
+        fn balance_of(self: @ContractState, account: ContractAddress) -> u256 {
+            assert(account.is_non_zero(), Errors::ADDRESS_ZERO);
+            self.balances.read(account)
+        }
+
+        fn owner_of(self: @ContractState, token_id: u256) -> ContractAddress {
+            let owner = self.owners.read(token_id);
+            assert(owner.is_non_zero(), Errors::INVALID_TOKEN_ID);
+            owner
+        }
+
+        fn get_approved(self: @ContractState, token_id: u256) -> ContractAddress {
+            assert(self._exists(token_id), Errors::INVALID_TOKEN_ID);
+            self.token_approvals.read(token_id)
+        }
+
+        fn is_approved_for_all(
+            self: @ContractState, owner: ContractAddress, operator: ContractAddress
+        ) -> bool {
+            self.operator_approvals.read((owner, operator))
+        }
+
+        fn approve(ref self: ContractState, to: ContractAddress, token_id: u256) {
+            let owner = self.owner_of(token_id);
+            assert(to != owner, Errors::APPROVAL_TO_CURRENT_OWNER);
+            assert(
+                get_caller_address() == owner
+                    || self.is_approved_for_all(owner, get_caller_address()),
+                Errors::NOT_TOKEN_OWNER
+            );
+            self.token_approvals.write(token_id, to);
+            self.emit(Approval { owner: self.owner_of(token_id), to: to, token_id: token_id });
+        }
+
+        fn set_approval_for_all(
+            ref self: ContractState, operator: ContractAddress, approved: bool
+        ) {
+            let owner = get_caller_address();
+            assert(owner != operator, Errors::APPROVE_TO_CALLER);
```

Comment:
This assert is unnecessary. Although it makes little sense, there's no harm in a wallet approving itself.